py/modbuiltins.c

```diff
@@ -137,30 +137,13 @@ MP_DEFINE_CONST_FUN_OBJ_1(mp_builtin_callable_obj, mp_builtin_callable);
 STATIC mp_obj_t mp_builtin_chr(mp_obj_t o_in) {
     #if MICROPY_PY_BUILTINS_STR_UNICODE
     mp_uint_t c = mp_obj_get_int(o_in);
-    uint8_t str[4];
-    int len = 0;
-    if (c < 0x80) {
-        *str = c;
-        len = 1;
-    } else if (c < 0x800) {
-        str[0] = (c >> 6) | 0xC0;
-        str[1] = (c & 0x3F) | 0x80;
-        len = 2;
-    } else if (c < 0x10000) {
-        str[0] = (c >> 12) | 0xE0;
-        str[1] = ((c >> 6) & 0x3F) | 0x80;
-        str[2] = (c & 0x3F) | 0x80;
-        len = 3;
-    } else if (c < 0x110000) {
-        str[0] = (c >> 18) | 0xF0;
-        str[1] = ((c >> 12) & 0x3F) | 0x80;
-        str[2] = ((c >> 6) & 0x3F) | 0x80;
-        str[3] = (c & 0x3F) | 0x80;
-        len = 4;
-    } else {
+    if (c >= 0x110000) {
         mp_raise_ValueError(MP_ERROR_TEXT("chr() arg not in range(0x110000)"));
     }
-    return mp_obj_new_str_via_qstr((char *)str, len);
+    vstr_t vstr;
+    vstr_init(&vstr, 0);
+    vstr_add_char(&vstr, c);
+    return mp_obj_new_str_from_vstr(&vstr);
```

Comment:
Thanks @jepler this is a good code size improvement and nice to reduce duplication!

The `vstr_t` here will be stack allocated, but the `vstr_add_char` will cause a one-block allocation. Previously this code was allocation-free (unless the qstr pool needs to grow).

Perhaps instead we could do:

```c
VSTR_FIXED(buf, 4);
vstr_add_char(&buf, c);
return mp_obj_new_str_via_qstr(buf->buf, buf->len);
```

That will also maintain the existing qstr interning behavior. We've discussed this before but I don't think we came to a satisfactory conclusion as there are advantages and disadvantages, but for now maintaining the behavior seems sensible (mainly to avoid allocations).

Reply:
Ah, I had tried a fixed vstr but it crashed in `mp_obj_new_str_from_vstr`, so I gave up on doing it free of allocation. I hadn't considered that `...via_qstr` must be OK with a stacked buffer.

I pushed this as a 2nd commit but can squash if preferred.

Reply:
Yeah mp_obj_new_str_from_vstr takes ownership of the vstr's buffer -- you cannot use it with fixed vstr.

Reply:
It's -112 bytes on stm32 (yay!) and tests are passing, this looks good! Thanks @jepler!

Reply:
care whether I squash these two commits?